qiskit/algorithms/optimizers/optimizer.py

```diff
@@ -12,15 +12,93 @@
 
 """Optimizer interface"""
 
-from typing import Dict, Any
+from typing import Dict, Any, Union, Callable, Optional, Tuple, List
 
+import warnings
 from enum import IntEnum
 import logging
 from abc import ABC, abstractmethod
 import numpy as np
 
+from qiskit.algorithms.algorithm_result import AlgorithmResult
+
 logger = logging.getLogger(__name__)
 
+POINT = Union[float, np.ndarray]
+
+
+class OptimizerResult(AlgorithmResult):
```

Comment:
Is this done for some sort of compatibility with the scipy OptimizerResult to allow scipy optimizers to be more directly used?

Reply:
The main motivation is to allow optimizers to return more information, e.g. to distinguish between gradient and function evaluations (which might be important since they differ in the circuits needed to evaluate) or to store additional data. But as a side effect it's close to SciPy's `OptimizeResult` so the algorithms can work with both result types.

Reply:
I was reading more than a 'side effect` hence the observation - it seemed intentional that the classes had same names. I was thinking you were going towards the idea of passing in a partial directly to a scipy optimizer and hence the result objects were compatible to allow that.